registry/handlers/layerupload.go

```diff
@@ -11,154 +11,124 @@ import (
+			httpErr = NewHTTPError(v2.ErrorCodeBlobUploadInvalid, err, http.StatusBadRequest)
+			return
 		}
 
-		if state.UUID != luh.UUID {
-			return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-				ctxu.GetLogger(ctx).Infof("mismatched uuid in upload state: %q != %q", state.UUID, luh.UUID)
-				w.WriteHeader(http.StatusBadRequest)
-				luh.Errors.Push(v2.ErrorCodeBlobUploadInvalid, err)
-			})
+		if state.UUID != uuid {
+			ctxu.GetLogger(ctx).Infof("mismatched uuid in upload state: %q != %q", state.UUID, uuid)
+			httpErr = NewHTTPError(v2.ErrorCodeBlobUploadInvalid, err, http.StatusBadRequest)
+			return
 		}
 
 		layers := ctx.Repository.Layers()
-		upload, err := layers.Resume(luh.UUID)
+		upload, err = layers.Resume(uuid)
 		if err != nil {
 			ctxu.GetLogger(ctx).Errorf("error resolving upload: %v", err)
 			if err == distribution.ErrLayerUploadUnknown {
-				return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-					w.WriteHeader(http.StatusNotFound)
-					luh.Errors.Push(v2.ErrorCodeBlobUploadUnknown, err)
-				})
+				httpErr = NewHTTPError(v2.ErrorCodeBlobUploadUnknown, err, http.StatusNotFound)
+				return
 			}
 
-			return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-				w.WriteHeader(http.StatusInternalServerError)
-				luh.Errors.Push(v2.ErrorCodeUnknown, err)
-			})
+			httpErr = NewHTTPError(v2.ErrorCodeUnknown, err, http.StatusInternalServerError)
+			return
 		}
-		luh.Upload = upload
 
 		if state.Offset > 0 {
 			// Seek the layer upload to the correct spot if it's non-zero.
 			// These error conditions should be rare and demonstrate really
 			// problems. We basically cancel the upload and tell the client to
 			// start over.
-			if nn, err := upload.Seek(luh.State.Offset, os.SEEK_SET); err != nil {
-				defer upload.Close()
+			if nn, err := upload.Seek(state.Offset, os.SEEK_SET); err != nil {
+				defer upload.Cancel() // Cancel calls Close
 				ctxu.GetLogger(ctx).Infof("error seeking layer upload: %v", err)
-				return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-					w.WriteHeader(http.StatusBadRequest)
-					luh.Errors.Push(v2.ErrorCodeBlobUploadInvalid, err)
-					upload.Cancel()
-				})
-			} else if nn != luh.State.Offset {
-				defer upload.Close()
-				ctxu.GetLogger(ctx).Infof("seek to wrong offest: %d != %d", nn, luh.State.Offset)
-				return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-					w.WriteHeader(http.StatusBadRequest)
-					luh.Errors.Push(v2.ErrorCodeBlobUploadInvalid, err)
-					upload.Cancel()
-				})
+				httpErr = NewHTTPError(v2.ErrorCodeBlobUploadInvalid, err, http.StatusBadRequest)
+				return
+			} else if nn != state.Offset {
+				defer upload.Cancel()
+				ctxu.GetLogger(ctx).Infof("seek to wrong offest: %d != %d", nn, state.Offset)
+				httpErr = NewHTTPError(v2.ErrorCodeBlobUploadInvalid, err, http.StatusBadRequest)
+				return
 			}
 		}
-
-		handler = closeResources(handler, luh.Upload)
```

Comment:
What closes the upload after a request now? I think I commented everywhere where the upload should be closed but please double check.